node/manager.go

```diff
@@ -0,0 +1,206 @@
+func (dm *DelegateManager) Stop(ctx context.Context) error {
+	if dm.broadcaster != nil {
+		return dm.broadcaster.Stop(ctx)
+	}
+	return nil
+}
+
+// HandleNodeInfo handle node info message
+func (dm *DelegateManager) HandleNodeInfo(ctx context.Context, peerID string, msg *iotextypes.ResponseNodeInfoMessage) {
+	log.L().Debug("delegate manager handle node info")
+	// recover pubkey
+	hash := hashNodeInfo(msg)
+	pubKey, err := crypto.RecoverPubkey(hash[:], msg.Signature)
+	if err != nil {
+		log.L().Warn("delegate manager recover pubkey failed", zap.Error(err))
+		return
+	}
+	// verify signature
+	if pubKey.Address().String() != msg.Info.Address {
+		log.L().Warn("delegate manager node info message verify failed", zap.String("address", msg.Info.Address))
+		return
+	}
+
+	dm.updateNode(&Info{
+		Version:   msg.Info.Version,
+		Height:    msg.Info.Height,
+		Timestamp: msg.Info.Timestamp.AsTime(),
+		Address:   msg.Info.Address,
+	})
+}
+
+// updateNode update node info
+func (dm *DelegateManager) updateNode(node *Info) {
+	addr := node.Address
+	dm.nodeMapMutex.Lock()
+	defer dm.nodeMapMutex.Unlock()
+
+	// update dm.nodeMap
+	dm.nodeMap[addr] = *node
+	// update metric
+	nodeDelegateHeightGauge.WithLabelValues(addr, node.Version).Set(float64(node.Height))
+}
+
+// BroadcastNodeInfo broadcast request node info message
+func (dm *DelegateManager) BroadcastNodeInfo(ctx context.Context) error {
+	log.L().Debug("delegate manager broadcast node info")
+	req, err := dm.genNodeInfoMsg()
+	if err != nil {
+		return err
+	}
+
+	// broadcast request meesage
+	if err := dm.transmitter.BroadcastOutbound(ctx, req); err != nil {
+		return err
+	}
+
+	// manually update self node info for broadcast message to myself will be ignored
+	dm.updateNode(&Info{
+		Version:   req.Info.Version,
+		Height:    req.Info.Height,
+		Timestamp: req.Info.Timestamp.AsTime(),
+		Address:   req.Info.Address,
+	})
+
+	return nil
+}
+
+// RequestSingleNodeInfoAsync unicast request node info message
+func (dm *DelegateManager) RequestSingleNodeInfoAsync(ctx context.Context, peer peer.AddrInfo) error {
+	log.L().Debug("delegate manager request one node info", zap.String("peer", peer.ID.Pretty()))
+
+	return dm.transmitter.UnicastOutbound(ctx, peer, &iotextypes.RequestNodeInfoMessage{})
+}
+
+// TellNodeInfo tell node info to peer
+func (dm *DelegateManager) TellNodeInfo(ctx context.Context, peer peer.AddrInfo) error {
+	log.L().Debug("delegate manager tell node info", zap.Any("peer", peer.ID.Pretty()))
+	req, err := dm.genNodeInfoMsg()
+	if err != nil {
+		return err
+	}
+
+	return dm.transmitter.UnicastOutbound(ctx, peer, req)
+}
+
+func (dm *DelegateManager) genNodeInfoMsg() (*iotextypes.ResponseNodeInfoMessage, error) {
+	req := &iotextypes.ResponseNodeInfoMessage{
+		Info: &iotextypes.NodeInfo{
+			Version:   version.PackageVersion,
+			Height:    dm.heightable.TipHeight(),
+			Timestamp: timestamppb.Now(),
+			Address:   dm.privKey.PublicKey().Address().String(),
```

Comment:
`Version` and `Address` won't change, so can save as `dm.version, dm.addr` at time of `New()` and directly used here

Reply:
good suggestion